### ISSUE TITLE
JPEG-compressed TIFF generated with VIPS fails to open in Photoshop due to invalid JPEG marker

### ISSUE THREAD

**Issue by dmolesUC** [libvips/libvips#2290]
## Bug report

When I try to use VIPS to convert a flat TIFF to a Pyramid TIFF with JPEG compression, the resulting image can't be opened in Photoshop due to an "invalid SOS JPEG marker".

(GitHub won't let me attach TIFFs to the issue, but the files can be found in repository [dmolesUC/vips-ps](https://github.com/dmolesUC/vips-ps) and are linked below. The example image is about 1.4 MB, and expands to about 2.3 MB after processing.)

**To Reproduce**

1. Given the image [`villette-040.tiff`](https://github.com/dmolesUC/vips-ps/raw/main/villette-040.tiff), execute the command:

   ```sh
   vips \
     tiffsave \
     villette-040.tiff \
     villette-040-p.tiff \
     --tile \
     --pyramid \
     --compression jpeg \
     --tile-width 256 \
     --tile-height 256
   ```
2. Open the resulting image [`villette-040-p.tiff`](https://github.com/dmolesUC/vips-ps/raw/main/villette-040-p.tiff) in Photoshop 2021.

**Expected behavior**

- Image opens.

**Actual behavior**

- Opening fails with

  > Could not open “villette-040-p.tiff” because an invalid SOS JPEG marker (Ss, Se, Ah, or Al) is found.

**Screenshots**

<img width="430" alt="Screen Shot 2021-06-08 at 11 00 56" src="https://user-images.githubusercontent.com/10374934/121236387-bed99e00-c84a-11eb-8640-3784d67a3da2.png">

**Environment**

 - OS: macOS 10.15.7
 - Vips: vips-8.10.6-Tue Mar 23 20:52:58 UTC 2021 (installed with Homebrew)

**Additional context**

The source image I've provided here is an uncompressed, 1-bit grayscale 2674x4358 image, but the same issue also occurs with an LZW-compressed 8-bit sRGB 8163x5587 source image.

**Comment by jcupitt** [comment]
Hi @dmolesUC,

Sorry, I don't have a copy of PS so it's hard for me to test. There's nothing odd about your TIFF, except that it's monochrome.

Does PS refuse to open all vips pyramids? What if you use a simple RGB JPEG as the source? Can you open `villette-040-p.tiff` in any other programs? You could try turning off the `--pyramid` or `--tile` flags and see if that changes things.

**Comment by dmolesUC** [comment]
I tried an RGB JPEG, with and without `--pyramid` & `--tile`, and got the same error (& so changed the issue title). I guess Photoshop doesn't like VIPS' JPEG compression? I've added a Photoshop-generated JPEG-compressed version for comparison, for whatever that's worth.

- [original JPEG](https://github.com/dmolesUC/vips-ps/raw/main/stgall-sm.jpg)
- [VIPS flat TIFF w/JPEG compression](https://github.com/dmolesUC/vips-ps/raw/main/stgall-sm-np.tif)
- [Photoshop flat TIFF w/JPEG compression](https://github.com/dmolesUC/vips-ps/raw/main/stgall-sm-ps.tif)

**Comment by dmolesUC** [comment]
I'm not any kind of image format person, but googling around, it looks like these markers (Se/Ss/Ah/Al) are used for progressive JPEGs, which may technically be disallowed by the [TIFF specs](http://libtiff.maptools.org/TIFFTechNote2.html).

> When the Compression field has the value 7, each image segment contains a complete JPEG datastream which is valid according to the ISO JPEG standard (ISO/IEC 10918-1).  Any sequential JPEG process can be used, including lossless JPEG, but progressive and hierarchical processes are not supported.

(7 is libtiff's [`COMPRESSION_JPEG`](https://gitlab.com/libtiff/libtiff/-/blob/master/libtiff/tiff.h#L184)

According to the [JPEG spec](https://www.w3.org/Graphics/JPEG/itu-t81.pdf) (see Table B.3 on page 38), it looks like SS/Se/Ah/Al should be set to 0, 63, 0, 0 for sequential JPEGs.

Is it possible that JPEG-compressed TIFFs are being written as progressive, or that when written as sequential, the wrong values are being set for these markers? I poked around in the code and didn't see anything obvious, but there are a lot of layers and I could easily have missed something. (Sorry I don't know the formats well enough to just look at the files and see what the values are — and also that Adobe didn't think to actually report or log them!)

**Comment by dmolesUC** [comment]
FWIW, if I use vips to generate either a sequential or progressive JPEG, Photoshop doesn't seem to have any troube with it.

**Comment by jcupitt** [comment]
Yes, I think it's libtiff. Your homebrew-written image won't load into QuPath correctly, but the same file created here loads fine.

I have a mac upstairs. I'll try updating homebrew and creating the image there.

**Comment by dmolesUC** [comment]
That opens fine! It looks like I have libtiff 4.3.0 and libjpeg 9d, and that those are the latest versions available in Homebrew.

**Comment by jcupitt** [comment]
Great!

9d is pretty old, but libtiff 4.3 is very bleeding edge, I expect that's the problem. I'm using ubuntu 21.04 and that's on libtiff 4.2.

I'm not sure what the best solution is. You probably need to edit your `tiff.rb` formula and rebuild. We should maybe open a bug report as well, probably with homebrew (downgrade libtiff as a workaround), and with libtiff (it obviously needs fixing).

**Comment by dmolesUC** [comment]
Yep, I just ran the same test in a Docker container running Alpine with libtiff 4.2.0 and it opens fine. That's where it's going to be running in production anyway, so I'm OK for now.

Thanks for helping me track this down. I'll file a libtiff issue.

**Comment by jcupitt** [comment]
OK, this should be fixed in 8.11.1. Please open a new issue if this is still broken in that version.
